text/0026-serialization.md

```diff
@@ -0,0 +1,242 @@
+          (byte)((value >> 32) & 0xff),
+          (byte)((value >> 40) & 0xff),
+          (byte)((value >> 48) & 0xff),
+          (byte)((value >> 56) & 0xff));
+  }
+  ```
+  
+  Each of call to `WriteByte*()` will have an additional call to `WriteCheckForSpace()`, which verifies there's enough space and resizes and copies the array if necessary, resulting in significant method call overhead and possible allocations. By comparison, FastBufferWriter aggressively inlines all of its writes, and writing any unmanaged type will always be done with a single assignment operation of that type followed by a single position increment (excepting special operations like packed writes that need to do transformation operations). Because we have very clear constraints on how much data can be sent to the transport for one message, we can make allocate buffers that fill those constraints and avoid any overhead of needing to resize them - we simply never will resize them. We also move our capacity checking to outside the write calls, so capacity can be checked just once for a series of several writes, dramatically reducing the overhead of capacity validation.
+  
+- **Mirror:**
+  
+  Mirror is not allocation-free - its NetworkReader and NetworkWriter classes must be instantiated. As a result, they do much the same that we do: they create PooledNetworkReader and PooledNetworkWriter classes and get and release them from a pool. By comparison, using mutable structs for FastBufferReader and FastBufferWriter involves no allocations at all, and they cannot be leaked.
+  
+  Mirror's NetworkReader and NetworkWriter are built around a managed byte[] array, and each write checks for capacity and grows the array if necessary, then performs copies using Array.ConstrainedCopy() when calling WriteBytes(). When calling other functions, such as WriteULong(), it writes one byte at a time as follows:
+  
+  ```c#
+  public static void WriteULong(this NetworkWriter writer, ulong value)
+  {
+      writer.WriteByte((byte)value);
+      writer.WriteByte((byte)(value >> 8));
+      writer.WriteByte((byte)(value >> 16));
+      writer.WriteByte((byte)(value >> 24));
+      writer.WriteByte((byte)(value >> 32));
+      writer.WriteByte((byte)(value >> 40));
+      writer.WriteByte((byte)(value >> 48));
+      writer.WriteByte((byte)(value >> 56));
+  }
+  ```
+  
+  Each of those calls to `WriteByte()` will have an additional call to `EnsureCapacity()`, which verifies there's enough space and resizes and copies the array if necessary, resulting in significant method call overhead and possible allocations. This is taken even further than UNet's implementation - where UNet would check for 8 bytes of space in a single check during `WriteByte8()`, Mirror checks for space before writing each and every individual byte. The same comments apply here as the ones I made discussing UNet.
+  
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
```

Comment:
The one thing I didn't see above but don't want to hold up the RFC for is mention of code generation.  Can I assume  there's nothing in the approach that would stand in the way of a code generation approach later?

Reply:
I personally am not a big fan of code gen for serialization... But that's just because I'm a control freak and don't like black boxes or things I can't tweak and optimize to the tiniest detail. But that's just my personal preference, and no, there's nothing that would stop us implementing it with this.